src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java

```diff
@@ -0,0 +1,691 @@
+            {
+                if (lineItemCoversInitialShardBounds((LineItem) entity))
+                {
+                    result = true;
+                }
+            }
+            else if (entity instanceof LocationItem)
+            {
+                if (locationItemCoversInitialShardBounds((LocationItem) entity))
+                {
+                    result = true;
+                }
+            }
+            else if (entity instanceof Relation)
+            {
+                result = relationMemberCoversInitialShardBoundsInternal(relation, (Relation) entity,
+                        parentRelationIdentifierTree);
+            }
+            else
+            {
+                throw new CoreException("Unknown Relation Member Type: {}",
+                        entity.getClass().getName());
+            }
+        }
+        return result;
+    }
+
+    private boolean relationMemberCoveredInternal(final Relation parentRelation,
+            final Relation relation, final Set<Long> parentRelationIdentifierTree)
+    {
+        boolean result = true;
+        final long newIdentifier = relation.getIdentifier();
+        if (parentRelationIdentifierTree.contains(newIdentifier))
+        {
+            logger.error(
+                    "Skipping! Unable to expand on relation which has a loop: {}. Parent tree: {}",
+                    parentRelation, parentRelationIdentifierTree);
+        }
+        else
+        {
+            final Set<Long> newParentRelationIdentifierTree = new HashSet<>();
+            newParentRelationIdentifierTree.addAll(parentRelationIdentifierTree);
+            newParentRelationIdentifierTree.add(newIdentifier);
+            if (!relationCoveredInternal(relation, newParentRelationIdentifierTree))
+            {
+                result = false;
+            }
+        }
+        return result;
+    }
+
+    private boolean relationMemberCoversInitialShardBoundsInternal(final Relation parentRelation,
+            final Relation relation, final Set<Long> parentRelationIdentifierTree)
+    {
+        boolean result = false;
+        final long newIdentifier = relation.getIdentifier();
+        if (parentRelationIdentifierTree.contains(newIdentifier))
+        {
+            logger.error(
+                    "Skipping! Unable to expand on relation which has a loop: {}. Parent tree: {}",
+                    parentRelation, parentRelationIdentifierTree);
+        }
+        else
+        {
+            final Set<Long> newParentRelationIdentifierTree = new HashSet<>();
+            newParentRelationIdentifierTree.addAll(parentRelationIdentifierTree);
+            newParentRelationIdentifierTree.add(newIdentifier);
+            if (relationCoversInitialShardBoundsInternal(relation, newParentRelationIdentifierTree))
+            {
+                result = true;
+            }
+        }
+        return result;
+    }
+
+    private boolean shouldBuildUnderlyingMultiAtlasWhenAddingNewShards(
```

Comment:
The logic here is the same, but the boolean operations have been broken up and commented out for clarity.